File: walking_robot_sim/src/quadropted_controller_cpp/src/odometry/odometry_state.cpp
#include <algorithm>

#include "quadropted_controller_cpp/odometry/odometry.hpp"

namespace quadropted {

OdometryState::OdometryState(int window)
    : filter_window_size(window), delta_x_queue(), delta_y_queue(), sum_delta_x(0.0), sum_delta_y(0.0) {
    // Инициализируем optional массивы
    for (int i = 0; i < 4; ++i) {
        prev_foot_positions[i] = std::nullopt;
    }
}

void OdometryState::append_delta(double dx, double dy) {
    if (static_cast<int>(delta_x_queue.size()) == filter_window_size) {
        sum_delta_x -= delta_x_queue.front();
        sum_delta_y -= delta_y_queue.front();
        delta_x_queue.pop_front();
        delta_y_queue.pop_front();
    }
    delta_x_queue.push_back(dx);
    delta_y_queue.push_back(dy);
    sum_delta_x += dx;
    sum_delta_y += dy;
}

std::pair<double, double> OdometryState::average_delta() const {
    int n = static_cast<int>(delta_x_queue.size());
    if (n == 0) return {0.0, 0.0};
    return {sum_delta_x / n, sum_delta_y / n};
}

void OdometryState::reset() {
    x = 0.0;
    y = 0.0;
    theta = 0.0;
    linear_velocity_x = 0.0;
    linear_velocity_y = 0.0;
    imu_angular_velocity = 0.0;
    delta_x_queue.clear();
    delta_y_queue.clear();
    sum_delta_x = 0.0;
    sum_delta_y = 0.0;
    for (int i = 0; i < 4; ++i) {
        foot_positions[i] = Eigen::Vector3d::Zero();
        prev_foot_positions[i] = std::nullopt;
        foot_contacts[i] = false;
    }
    joint_positions.fill(0.0);
    gazebo_clock_sec = 0;
    gazebo_clock_nanosec = 0;
    encoder_pos = 0;
}

}  // namespace quadropted

File: walking_robot_sim/src/quadropted_controller/scripts/QuadrupedOdometry/odometry_state.py
#!/usr/bin/env python3
"""
Состояние одометрии четвероногого робота.
Вынесено из QuadrupedOdometryNode.py при декомпозиции.
"""

from collections import deque


class OdometryState:
    """Чистое состояние одометрии (без ROS зависимостей)."""

    def __init__(self, filter_window_size=14):
        # Позиция и ориентация
        self.x = 0.0
        self.y = 0.0
        self.theta = 0.0

        # Скорости
        self.linear_velocity_x = 0.0
        self.linear_velocity_y = 0.0
        self.imu_angular_velocity = 0.0

        # Фильтр скользящего среднего
        self.filter_window_size = filter_window_size
        self.delta_x_queue = deque(maxlen=filter_window_size)
        self.delta_y_queue = deque(maxlen=filter_window_size)

        # Позиции лап
        self.prev_foot_positions = [None, None, None, None]  # [FR, FL, RR, RL]
        self.foot_positions = [(0.0, 0.0, 0.0)] * 4

        # Контакты и суставы
        self.foot_contacts = [False, False, False, False]
        self.joint_positions = [0.0] * 12

        # Внешние данные
        self.gazebo_clock_sec = 0
        self.gazebo_clock_nanosec = 0
        self.encoder_pos = 0

    def reset(self):
        """Сбросить состояние к начальным значениям."""
        self.x = 0.0
        self.y = 0.0
        self.theta = 0.0
        self.linear_velocity_x = 0.0
        self.linear_velocity_y = 0.0
        self.imu_angular_velocity = 0.0
        self.delta_x_queue.clear()
        self.delta_y_queue.clear()
        self.prev_foot_positions = [None, None, None, None]
        self.foot_positions = [(0.0, 0.0, 0.0)] * 4
        self.foot_contacts = [False, False, False, False]
        self.joint_positions = [0.0] * 12
